subprojects/language-java/src/main/java/org/gradle/api/tasks/javadoc/internal/JavadocToolAdapter.java

```diff
@@ -19,12 +19,12 @@
 import org.gradle.api.file.RegularFile;
 import org.gradle.api.tasks.WorkResult;
 import org.gradle.jvm.toolchain.JavaInstallationMetadata;
-import org.gradle.jvm.toolchain.JavadocTool;
-import org.gradle.jvm.toolchain.internal.DefaultJavaToolchainUsageProgressDetails.JavaTool;
+import org.gradle.jvm.toolchain.internal.DefaultJavaToolchainUsageProgressDetails;
 import org.gradle.jvm.toolchain.internal.JavaToolchain;
+import org.gradle.jvm.toolchain.internal.JavadocToolInternal;
 import org.gradle.process.internal.ExecActionFactory;
 
-public class JavadocToolAdapter implements JavadocTool {
+public class JavadocToolAdapter implements JavadocToolInternal {
```

Comment:
I think we shouldn't use those `*Internal` interfaces. See my changes in #21491

https://github.com/gradle/gradle/pull/21491/commits/3f62209145744b45cf24ab998d694e4b03f9ad39

Reply:
I guess adding a side effect on the toolchain provider (and not the tool provider directly) is fine as long as we do in the same place, like you suggest here:
https://github.com/gradle/gradle/blob/3f62209145744b45cf24ab998d694e4b03f9ad39/subprojects/platform-jvm/src/main/java/org/gradle/jvm/toolchain/internal/JavaToolchainQueryService.java#L67-L69

I didn't do it initially, because it was not clear how the side effect are propagated. But it seems to work as expected 👍